File: Whattowatch/app/src/main/java/com/simranjeetsingh05/whattowatch/activities/MainActivity.java
package com.simranjeetsingh05.whattowatch.activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;

import com.simranjeetsingh05.whattowatch.R;
import com.simranjeetsingh05.whattowatch.adapters.TVShowsAdapter;
import com.simranjeetsingh05.whattowatch.databinding.ActivityMainBinding;
import com.simranjeetsingh05.whattowatch.listeners.TVShowsListener;
import com.simranjeetsingh05.whattowatch.models.TvShow;
import com.simranjeetsingh05.whattowatch.viewmodels.MostPopularTVShowsViewModel;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements TVShowsListener {

    private ActivityMainBinding activityMainBinding;
    private MostPopularTVShowsViewModel viewModel;
    private List<TvShow> tvShows = new ArrayList<>();
    private TVShowsAdapter tvShowsAdapter;
    private int currentPage = 1;
    private int totalAvailablePages = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityMainBinding = DataBindingUtil.setContentView(this,R.layout.activity_main);
        doInitialization();
    }


    private  void doInitialization(){
        activityMainBinding.tvshowsRecyclerView.setHasFixedSize(true);
        viewModel  = new ViewModelProvider(this).get(MostPopularTVShowsViewModel.class);
        tvShowsAdapter = new TVShowsAdapter(tvShows, this);
        activityMainBinding.tvshowsRecyclerView.setAdapter(tvShowsAdapter);
        activityMainBinding.tvshowsRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if(!activityMainBinding.tvshowsRecyclerView.canScrollVertically(1)){
                    if (currentPage <= totalAvailablePages){
                        currentPage += 1;
                        getMostPopularTVShows();
                    }
                }
            }
        });
        getMostPopularTVShows();
    }
    private void getMostPopularTVShows(){
        toggleLoading();
        viewModel.getMostPopularTVShows(currentPage).observe(this, mostPopularTVShowsResponses ->{
                    toggleLoading();
                    if(mostPopularTVShowsResponses != null){
                        totalAvailablePages = mostPopularTVShowsResponses.getPages();
                        if(mostPopularTVShowsResponses.getTvShows() != null){
                            int oldCount = tvShows.size();
                            tvShows.addAll((mostPopularTVShowsResponses.getTvShows()));
                            tvShowsAdapter.notifyItemRangeInserted(oldCount, tvShows.size());

                        }
                    }
                });
    }


    private  void toggleLoading() {
        if(currentPage == 1){
            if(activityMainBinding.getIsLoading() != null && activityMainBinding.getIsLoading()){
                activityMainBinding.setIsLoading(false);
            }else{
                activityMainBinding.setIsLoading(true);
            }
        }else{
            if(activityMainBinding.getIsLoading() != null && activityMainBinding.getIsLoading()){
                activityMainBinding.setIsLoading(false);
            }else{
                activityMainBinding.setIsLoading(true);
            }


        }
    }

    @Override
    public void onTVShowClicked(TvShow tvShow) {
        Intent intent = new Intent(getApplicationContext(), TVShowDetailsActivity.class);
        intent.putExtra("id",tvShow.getId());
        intent.putExtra("name",tvShow.getName());
        intent.putExtra("startDate",tvShow.getStartDate());
        intent.putExtra("country",tvShow.getCountry());
        intent.putExtra("network",tvShow.getNetwork());
        intent.putExtra("status",tvShow.getStatus());
        startActivity(intent);
    }
}

File: AndroidKotlin/app/src/main/java/com/simranjeetsingh05/EmailContacts/activities/MainActivity.kt
package com.simranjeetsingh05.EmailContacts.activities

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.simranjeetsingh05.EmailContacts.R
import com.simranjeetsingh05.EmailContacts.adapters.MyRecyclerViewAdapter
import com.simranjeetsingh05.EmailContacts.database.People
import com.simranjeetsingh05.EmailContacts.database.PeopleDatabase
import com.simranjeetsingh05.EmailContacts.databinding.ActivityMainBinding
import com.simranjeetsingh05.EmailContacts.repository.PeopleRepository
import com.simranjeetsingh05.EmailContacts.viewmodel.PeopleViewModel
import com.simranjeetsingh05.EmailContacts.viewmodel.PeopleViewModelFactory

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private lateinit var peopleViewModel: PeopleViewModel
    private lateinit var adapter: MyRecyclerViewAdapter
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        setContentView(R.layout.activity_main)
        binding = DataBindingUtil.setContentView(this,R.layout.activity_main)
        val dao = PeopleDatabase.getInstance(application).peopleDAO
        val repository = PeopleRepository(dao)
        val factory = PeopleViewModelFactory(repository)
        peopleViewModel = ViewModelProvider(this,factory).get(PeopleViewModel::class.java)
        binding.myViewModel = peopleViewModel
        binding.lifecycleOwner = this
        initRecyclerView()

        peopleViewModel.message.observe(this, Observer {
            it.getContentIfNotHandled()?.let{
                Toast.makeText(this,it,Toast.LENGTH_LONG).show()
            }
        })

    }
    private fun initRecyclerView(){
        binding.peopleRecyclerView.layoutManager = LinearLayoutManager(this)
        adapter = MyRecyclerViewAdapter({ selectedItem: People -> listItemClicked(selectedItem) })
        binding.peopleRecyclerView.adapter = adapter
        displayPeopleList()

    }

    private fun displayPeopleList(){
        peopleViewModel.people.observe(this, Observer {
            Log.i("MYTAG",it.toString())
            adapter.setList(it)
            adapter.notifyDataSetChanged()
        })
    }

    private fun listItemClicked(people: People){
//        Toast.makeText(this,"Selected name is ${people.name}",Toast.LENGTH_LONG).show()
        peopleViewModel.initUpdateAndDelete((people))

    }
}